main.cpp

```diff
@@ -2205,34 +2360,45 @@ struct picoboot_memory_access : public memory_access {
         vector<uint8_t> write_data; // used when erasing flash
         if (flash == get_memory_type(address, model)) {
             connection.exit_xip();
-            if (erase) {
-                // Do automatically erase flash, and make it aligned
-                // we have to erase in whole pages
-                range aligned_range(address & ~(FLASH_SECTOR_ERASE_SIZE - 1),
-                                    ((address + size) & ~(FLASH_SECTOR_ERASE_SIZE - 1)) + FLASH_SECTOR_ERASE_SIZE);
-                assert(aligned_range.contains(address));
-                assert(aligned_range.contains(address + size));
-
-                uint32_t pre_len = address - aligned_range.from;
-                uint32_t post_len = aligned_range.to - (address + size);
-                assert(pre_len + size + post_len == aligned_range.len());
-
-                // save data before the changing data
-                write_data.resize(pre_len);
-                read(aligned_range.from, write_data.data(), write_data.size(), false);
-                // now add the data that is changing
-                write_data.insert(write_data.end(), buffer, buffer + size);
-                // save data after the changing data
-                write_data.resize(aligned_range.len());
-                read(address + size, write_data.data() + pre_len + size, post_len, false);
-
-                // Do the erase
-                connection.flash_erase(aligned_range.from, aligned_range.len());
-
-                // Update what will now be written
-                address = aligned_range.from;
-                buffer = write_data.data();
-                size = aligned_range.len();
+            // Flash Translation Layer - auto-erase, and only write changed data
+            if (enable_ftl) {
+                // Check what's there already
+                write_data.resize(size);
+                read(address, write_data.data(), size, false);
+                // Check if we even need to write
+                if (std::equal(write_data.cbegin(), write_data.cend(), buffer)) {
+                    return;
+                }
+                // Check if we need to erase (ie check for non 0xff)
+                if (!std::all_of(write_data.cbegin(), write_data.cend(), [](uint8_t v) { return v == 0xff; })) {
+                    // Do automatically erase flash, and make it aligned
```

Comment:
```suggestion
                    // To automatically erase flash, and make it aligned
```
?

Reply:
The comments were meant to mean “Check if we want to erase” followed by “Yes, we do want erase” - so maybe just delete the “Do”?
```suggestion
                    // Automatically erase flash, and make it aligned
```

Reply:
Ahhh, I didn't realise that
```c
// Do automatically erase flash, and make it aligned
// we have to erase in whole pages
```
was intended to be two separate sentences, and not one long sentence :slightly_smiling_face: